### PR TITLE
Give access to FilterChain list of loaded filters

### PR DESCRIPTION
**Pull request by peci1**
This PR makes `FilterChain` members `reference_pointers_`, `buffer0_` and `buffer1_` protected instead of private. This helps writing custom filter chain implementations that e.g. need to access the list of loaded filters or that do updating in a different way and want to utilize the already existing buffer variables.

If making the buffers protected would seem undesirable, that part of the PR can be reverted. The buffers are really implementation-specific things and making them protected would just allow child classes to utilize them the same way `FilterChain` does. But I don't think it hurts exposing these, too (they're only used in the `update()` method and serve as cache variables with no meaning outside the method; moreover, the `update()` method wipes any content of the buffers when it is called, so even if the buffers contain some unexpected values, it should not break the working of the `update()` metod). The only thing that would hurt the operation of the chain would be asynchronous calls, but as `update()` isn't thread-safe at all even now, I wouldn't see that as a problem.

### CODE REVIEW

**Comment by jonbinney** [comment]
I just took a quick look at this; my first thought is that not all child classes would need these data structures - they may want to handle these buffers in different ways. If that is true, shouldn't we just let each of them create whatever data structures they want?

**Comment by peci1** [comment]
Yes, the buffers are for discussion. But `reference_pointers_` are hard to substitute in child classes.

**Comment by jonbinney** [comment]
Do you think that child classes will want to modify `reference_pointers_`, or just use them to implement their own `update()` function? If they only need read access, a protected `FilterChain::get_filters()`  could return a const ref to the vector of Filters.

**Comment by peci1** [comment]
`reference_pointers_` is a vector to non-const shared_ptrs to non-const filters, so the potential `get_filters()` method returning a const ref would just mean the downstream packages cannot add/remove filters, but they could change the pointers. I don't see a reason for this limitation. Generally, when using protected members/methods, the downstream package should very well be aware of what it can and can't do.

**Comment by peci1** [comment]
But generally, adding a protected function for accessing the private member would also be an option (if that's preferred).

**Comment by jonbinney** [comment]
Sorry it has been taking me such a long time to reply on this. I do worry about the fragility of having a protected member variable that child classes in projects outside of this repo will use. If the type or name of `reference_pointers_` is ever changed in this repo for any reason, it would break other people's code and we'd have no way of knowing. To make things worse, noetic is the final ROS1 distro, so that change couldn't wait for a new ROS distro. People's robot's would just stop working, and their code would stop compiling.

That said, your use case is valid and I think it is worth doing. I can't think of any better way of doing it. I'd like to see a few things done before merging this to mitigate the risk of problems:
- rename "`reference_pointers_`" to "`filters_`". `reference_pointers_` is a terrible name and we shouldn't change it later since it could break child classes.
- change the type of the vector to use `std::shared_ptr` instead of `boost::shared_ptr`. Most other ROS code has changed over.
- leave the buffers as private, to minimize the member variables touched by child classes.

Does that sound ok to you?

**Comment by peci1** [comment]
I understand your concerns about forward- and backward-compatibility. However, I think your proposal is even more breaking.

If there is existing code accessing this member via a hack (`#define private protected`), it would stop compiling. I know such code is deemed to be broken if it touches private members, but breaking it without a good reason is bad from my point of view. I have personally written some internal code accessing the private member (that's why I sent this PR - to make the code less hacky and prone to break).

Similarly for the `boost::shared_ptr`->`std::shared_ptr`. Boost dependency will probably not be removed by the proposed change, so why changing it at all?

So my proposal would be to add a non-virtual accessor method that returns a reference to `reference_pointers_` member. The name could be `getFilters()` to partially fix the bad naming of the private variable. I'm just not sure whether this method should return a const reference or a mutable one. But I think const makes more sense as it could be dangerous to allow downstream code to modify the list of filters.

What do you think about this other proposal?

**Comment by jonbinney** [comment]
I like the accessor with a const reference but I feel pretty strongly about `std::shared_ptr`. In my opinion, breaking hacked code is not as bad as making it impractical to ever update the code to move to the shared pointers which are in the standard library. Yes, hacked code would need to be updated once, but after that the variable can be reliably used.

I thought of one other possibility which might be provide the needed functionality. A non-member function called `load_filters()` could take a node handle as a parameter, load the filters based on the ROS parameters, and return the vector of filters. The main filter chain class could call this function to load its filters, as could any other class that wants to use filters. Since (I assume) people would want to write their own `update()` function, inheritance wouldn't be providing anything at that point. No member variables would need to be shared.

I'd still want that function to return a vector of `std::shared_ptr` though. The only other boost dependency on boost in this package is one variable in the realtime circular buffer, and it is private and could potentially be removed in the future.

**Comment by peci1** [comment]
> breaking hacked code is not as bad

I agree, hacky code should not hold back development. I just wanted to say "if it doesn't need changing, then don't change it".

If you as a maintainer feel the need to change the pointer type, then it's probably a good enough reason to do the change (by the way, I see only Tully in the maintainer list in package.xml...).

Regarding the `load_filters` function - I think it would be mostly equivalent to creating a `filters::FilterChain` and reading/copying the filters list (if you don't want to inherit from it).

So I'm inclined to go with:
1. rename `reference_pointers_` to `filters_` and change to `std::shared_ptr`
2. add a const ref getter `getFilters()` returning `filters_`

Would that be okay?

**Comment by jonbinney** [comment]
Oops, looks like I forgot to update all branches when I got added as a maintainer. Only updated the ros2 branch: https://github.com/ros/filters/blob/5507bc5175cb74e825ee20b9dc19e67ef6a4a022/package.xml#L11

Since this is a change to the external API, it would be good to get tully's opinion too. @tfoote What do you think about the proposals here? To summarize, it @peci1 has pointed out that it would be useful for people to be able to create their own variants of the FilterChain class without needing to reimplement everything. There are a few options discussed above, but the main two at this point are:
1. Create a `FilterChain::getFilters()` function that returns a const reference to the filters vector. Other classes can then use them to implement their own `update()`
2. Create a non-member function called `loadFilters()ros::NodeHandle node_handle)` that loads the filters based on ros params. This would be used by the FilterChain class and by anyone else.

For either of these, we could either return a vector of `boost::shared_ptr` of filters or update to `std::shared_ptr`. At the moment, no public member variables, functions, or member functions in the filters repo return boost types. But as @peci1 points out, the private vector of filters is currently using `boost::shared_ptr` and we could leave that alone.

**Comment by tfoote** [comment]
We should definitely be forward looking and use `std::` instead of `boost::` in any public APIs. We worked hard to remove boost from any of our core implmentations of ROS 2 so we can keep our dependency footprint small. 

Part of my hope for ROS 2 was that the need for filters would go away. But we haven't gotten to that point of having the feature set for that yet. 

Providing the accessor method `getFilters()` for the list of filters makes sense to me. It should definitely return a `std::vector` type though not a boost one. Deboostifying the implementation can be a background task. But we should make sure not to make that harder. 

And the function to load from parameters would be potentially helpful. But it's probably worth separating from adding the accessor first.

**Comment by peci1** [comment]
Thanks @tfoote for the comment. I've force-pushed a new version that adds the `getFilters()` function and changes the internal pointers to std pointers.

I needed to get unmanaged pointers from pluginlib because it can only make boost pointers, but as I wrap them in a std::shared_ptr almost immediately, there should be no problem with dangling pointers.

File: include/filters/filter_chain.hpp
Comment (tfoote) [inline include/filters/filter_chain.hpp]: For lifetimes here. Does it make sense to pass a reference to the internal list. Or to create a new vector of shared pointers that can be returned and the shared pointers can be trusted to maintain validity. I guess in this case if you want the shared ptr target to persist you can just make a copy from the one in this const vector reference.

**Review by tfoote — APPROVED** [review]
Overall this looks reasonable to me. I'll jet Jon have the final signoff though. Also we need to confirm what branches onto which we should land this.

File: include/filters/filter_chain.hpp
Comment (peci1) [reply include/filters/filter_chain.hpp]: I don't see a problem with returning a reference to the internal list. If the users expects his code might outlive the filterchain, making a copy of the list is easy. The usual use-case as I see it is iterating over loaded filters in FilterChain descendant classes, which would probably want to access the actual list and not its (possibly outdated) copy.

Comment (tfoote) [reply include/filters/filter_chain.hpp]: If the vector is possibly able to mutate we definitely should be returning a copy. std::vector is not threadsafe.

File: include/filters/filter_chain.hpp
Comment (jonbinney) [inline include/filters/filter_chain.hpp:483]: Could you add a comment here explaining what you said in the PR? (that we create an unmanaged plugin first because pluginlib doesn't have a version that returns `std::shared_ptr`.

File: test/test_chain.cpp
Comment (jonbinney) [inline test/test_chain.cpp:114]: Thanks for adding a test!

File: include/filters/filter_chain.hpp
Comment (jonbinney) [inline include/filters/filter_chain.hpp:235]: Could you add a comment here explaining what you said in the PR? (that we create an unmanaged plugin first because pluginlib doesn't have a version that returns std::shared_ptr.

**Review by jonbinney — CHANGES_REQUESTED** [review]
Code looks good except for changing `getFilters` to return a copy instead of a reference. I requested a couple of comments to help future us.

**Comment by peci1** [comment]
I've added the comments and changed `getFilters()` to return a copy of the internal list of filters.

For good usability in downstream code, I've also added `numFilters()` and `isConfigured()` functions so that the downstream code can easily recognize when its copy of the filter list went out of date. The number of filters should only increase when calling `configure()` and can be reset to zero by calling `clear()`. So I think a combination of checking the number of filters and the configured flag should be enough to tell whether an update of the filter list is needed or not (theoretically, the chain can be cleared and then reloaded with the same number of filters, so the client code could be tricked to not update its copy, but I guess this is a marginal case which can be resolved e.g. by some periodical update of the copy on a low frequency).

**Comment by jonbinney** [comment]
Rather than checking numFilters and isConfigured, wouldn't it be simpler for client code to always call getFilters before doing anything with them?

**Comment by peci1** [comment]
Imagine an alternative implementation of `update()`. You want to iterate through all filters in each call. Copying the vector of filters each time doesn't sound like the best idea... (I'm referring to an actual use in code I maintain - there is a dynamic reconfigure call that can disable each single filter in the chain).

**Comment by jonbinney** [comment]
"copying the vector" should be pretty low overhead considering it's just a few shared pointers. Unless you're using this in hard realtime code, i doubt the performance hit will be noticeable. On the other hand, I'm pretty sure laser_filters in general isn't intended for us in hard realtime.

Do you have any profiling data to show that the performance hit is substantial? Otherwise keeping the API simple seems more important.

**Comment by peci1** [comment]
Performance testing code:

<details>

```c++
std::vector<std::shared_ptr<int>> vec(10);
  
  for (size_t j = 0; j < 10; ++j)
  {
    auto start = ros::WallTime::now();
    int* addr = 0;
    {
      for (size_t i = 0; i < 1000; ++i)
      {
        auto& vec2 = vec;
        for (const auto& ii: vec2)
          addr = reinterpret_cast<int*>(addr - ii.get());
      }
    }
    auto end = ros::WallTime::now();
    ROS_WARN("Ref took  %f s, result is %lu", (end - start).toSec(), (unsigned long) addr);

    start = ros::WallTime::now();
    addr = 0;
    {
      for (size_t i = 0; i < 1000; ++i)
      {
        auto vec2 = vec;
        for (const auto& ii: vec2)
          addr = reinterpret_cast<int*>(addr - ii.get());
      }
    }
    end = ros::WallTime::now();
    ROS_WARN("Copy took %f s, result is %lu", (end - start).toSec(), (unsigned long) addr);
  }
```

</details>

Results on my notebook (I only show one of the 10 iterations as all look basically the same):

Release mode:

```
Ref took  0.000016 s, result is 0
Copy took 0.000061 s, result is 0
```

Debug mode:

```
Ref took  0.000201 s, result is 0
Copy took 0.000973 s, result is 0
```

So in 1000 runs of a filterchain containing 10 filters, copying the vector makes the code 50 us slower in release mode and 700 us slower in debug mode. That is about 50 ns per filterchain run in release mode.

I agree 50 ns is not much, but this is on a pretty fast Ryzen CPU. Still, if I can decide whether to save 50 ns or not, I would say generally yes.

---

> Otherwise keeping the API simple seems more important.

The `numFilters()` and `isConfigured()` do not seem to me like they are complicating the API. They're just accessors to protected members (or the size of them). Nobody is forced to use these methods. And somebody might find them useful.

However, if you're firm in your opinion that these methods should not be a part of this PR, I'll remove them to make this going.

**Comment by jonbinney** [comment]
@pec1 thanks for doing that profiling! I have very little intuition about how fast something like this is in practice.

I agree that faster is better than slower, but in this case I think that the slower option is "fast enough" and it results in a much simpler API. Lets stick with just the single getFilters function.

**Comment by peci1** [comment]
Okay. I left just the comments and a few added pieces of tests that test whether the returned vector is really a copy and not a reference to the internal vector.

**Review by jonbinney — APPROVED** [review]

**Comment by jonbinney** [comment]
Looks good! Thanks @peci1 !

**Comment by peci1** [comment]
Great! Thank you too, Jon. Would you please take care about backporting to Melodic?

**Comment by jonbinney** [comment]
Backported in this PR: https://github.com/ros/filters/pull/61 . The filename changed from filter_chain.h in lunar-devel to filter_chain.hpp in noetic-devel, so I squashed the commits in this PR and then manually applied the squashed patch to that file. @peci1 could you take a look and make sure I didn't do anything stupid?

**Comment by peci1** [comment]
Thanks a lot, @jonbinney ! Could I ask you for new releases for both Melodic and Noetic? It seems for Melodic, this is the only change after 1.8.1, and for Noetic, there is #58 in addition to this PR.

**Comment by jonbinney** [comment]
@peci1 sure thing.
Melodic release: https://github.com/ros/rosdistro/pull/33120
Noetic release: https://github.com/ros/rosdistro/pull/33121

**Pull request merged**
